### PR TITLE
[docs] VACMS-3157 Update code review document

### PR DESCRIPTION
**Pull request by indytechcook**
## Description

See #3157 

This document assumes that tugboat has been implemented for PR tests.

### CMS user-facing annoucement

Is an announcement needed to let editors know of this change? 
- [ ] Yes, and it's written in issue ____ and queued for publication. 
  - [ ] Merge and ping @ rachel-kauff so she's ready to publish after deployment
- [ ] Yes, but it hasn't yet been written 
  - [ ] Don't merge yet -- ping @ rachel-kauff to prompt her to write and queue content
- [x] No announcement is needed for this code change. 
  - [x] Merge & carry on unburdened by announcements

---
This change is [Reviewable](https://reviewable.io/reviews/department-of-veterans-affairs/va.gov-cms/3437)

### CODE REVIEW

File: READMES/code-review.md [hunk 1]
```diff
@@ -1,15 +1,99 @@
...
+1. PR Reviewer will provide feedback on the Pull Request using the Review Changes feature. The feedback should be in a single review and not several individual comments on the Pull Request.
+1. The PR Owner will provide changes requested by the Code Review and answer all questions.
+1. When PR Reviewer is satisfied the changes introduced by the Pull Request are of high quality, address the related issue and will not introduce new regressions, the Pull Request is approved.
+1. The PR Owner merges the Pull Request.
+
+## Automated Code Qualtiy Review Tools
+
+The following automated code tools must pass before a Pull Request can be merged.
+
+* PHPCS
```
Comment (swirtSJW) [inline READMES/code-review.md:89]: If these are part of our test suite, do they need to be called out separately?

**Review by swirtSJW — APPROVED** [review]

**Review by ndouglas — DISMISSED** [review]

**Comment by ndouglas** [comment]
One question: why is it important for the PR to be assigned to the author?  I'm guessing that serves some function, I'm just not sure what.  And maybe that should be in the document? 🤔

File: READMES/code-review.md [hunk 2]
```diff
@@ -1,15 +1,99 @@
...
 
-One (or more) lead developers is assigned to take the responsibility of merging all pull requests. This ensures consistency in quality control as well as identifying any potential issues with related, open pull requests.  GIthub is configured to not allow merging a Pull Request until it has been reviewed and approved.
+## Guiding principles
 
-A small group of one or more person(s) is selected to be integrators. All commits are reviewed by this group. If work is done by an integrator, their work should be reviewed by a fellow integrator (As if they were a developer.  You can't review your own code.).
+* Main/Master branch must always deployable
+* Tests must pass 
+* The cycle time from Pull Request open to Pull Request merged should be minimized
+* PR Owners must be empowered and trusted to merge code
+* PR Review must Approve Pull Request
```
Comment (oksana-c) [inline READMES/code-review.md:9]:
```suggestion
* PR Reviewer must Approve Pull Request
```

File: READMES/code-review.md [hunk 3]
```diff
@@ -1,15 +1,99 @@
...
+* Automated Test Suite must be trustworthy and inspire confidence
+
+## Personas
+
+Let's start the discussion by talking about different personas. I will try to avoid process descriptions in the personas and focus on responsibilities.
```
Comment (oksana-c) [inline READMES/code-review.md:14]:
```suggestion
Let's start the discussion by talking about different personas and focus on their responsibilities.
```

File: READMES/code-review.md [hunk 4]
```diff
@@ -1,15 +1,99 @@
...
+The PR Owner will be the person `assigned` to the pull request.
+
+*Code Owner Responsibilities:*
+
+* All changes introduced within the Pull Request
+* Responding to questions and requested changes to the Pull Request
+* Passing tests
+* Adding new tests
+* Adding documentation
+* Merging the Pull Request on Approval and test passing.
```
Comment (oksana-c) [inline READMES/code-review.md:39]:
```suggestion
* Merging the Pull Request when it's approved, tests are passing and the PR meets a Definition of Done.
```

File: READMES/code-review.md [hunk 5]
```diff
@@ -1,15 +1,99 @@
...
+* Product Manager
+
+While the PR Owner has ownership of the changes introduced in the Pull Request, both the PR Owner and PR Reviewer take shared responsibility for the quality of the changes.
+
+### PR Owner
+
+A PR Owner is either the author of the code, the creator of the Pull Request, or another designated person. There should be only a single PR Owner per Pull Request. A PR Owner can be anyone with a Github account.
+
+_Merging the Pull Request on Approval and test passing needs to read Merging the Pull Request when it's approved, tests are passing and the PR meets a Definition of Done._
+
```
Comment (oksana-c) [inline READMES/code-review.md:29]:
```suggestion
```

File: READMES/code-review.md [hunk 6]
```diff
@@ -1,15 +1,99 @@
...
+*Code Owner Responsibilities:*
+
+* All changes introduced within the Pull Request
+* Responding to questions and requested changes to the Pull Request
+* Passing tests
+* Adding new tests
+* Adding documentation
+* Merging the Pull Request on Approval and test passing.
+* Validating the changes in the Pull Request fulfill the needs of the Product Owner and related issue
```
Comment (oksana-c) [inline READMES/code-review.md:40]:
```suggestion
* Validating the changes in the Pull Request fulfill the needs of the Product Owner and ACs in related issue
```

File: READMES/code-review.md [hunk 7]
```diff
@@ -1,15 +1,99 @@
...
+* Communicating any status about the deploy process, extra downtime, etc to the appropriate team members and Product Owner.
+* Provide notes on how to test and any necessary screenshots of changes.
+* Request a review from a Tech Lead if necessary.
+* Request a review from a Product Owner on CI if necessary.
+
+### PR Reviewer
+
+A PR Review is one or more people responsible for reviewing all the changes a Pull Request introduces. There can be multiple PR Reviewers per PR. A PR Reviewer can be anyone that is part of the department-of-veterans-affairs/va.gov-cms group. If a change introduced by a Pull Request introduces an architectural change or a significant new feature, a Tech Lead should be the PR Reviewer.
```
Comment (oksana-c) [inline READMES/code-review.md:48]:
```suggestion
A PR Review is one or more people responsible for reviewing all the changes a Pull Request introduces. There can be multiple PR Reviewers per PR. A PR Reviewer can be anyone that is part of the department-of-veterans-affairs/va.gov-cms group. If a change introduced by a Pull Request introduces an architectural change or a significant new feature, a Tech Lead and a Product Manager (as needed) should be the PR Reviewer.
```

File: READMES/code-review.md [hunk 8]
```diff
@@ -1,15 +1,99 @@
...
+
+_Merging the Pull Request on Approval and test passing needs to read Merging the Pull Request when it's approved, tests are passing and the PR meets a Definition of Done._
+
```
Comment (oksana-c) [inline READMES/code-review.md:51]:
```suggestion
```

File: READMES/code-review.md [hunk 9]
```diff
@@ -1,15 +1,99 @@
...
+* A PR Reviewer will self assign or be assigned by the PR Owner as a Reviewer on the Pull Request.
+* PR Reviewer Responsibilities
+* Review code for quality passed upon Insert Spec Here?
```
Comment (oksana-c) [inline READMES/code-review.md:54]:
```suggestion
* Review code for quality passed upon [Insert Spec Here]
```

File: READMES/code-review.md [hunk 10]
```diff
@@ -1,15 +1,99 @@
...
+* Test Code changes on CI
+* Document what was tested providing screenshots if necessary.
+* Follow up with questions/feedback PR Code Owner in a timely manner
+* Approve Pull Request or provide detailed feedback on why the Pull Request was not approved.
+* Request a review from a Tech Lead if necessary.
```
Comment (oksana-c) [inline READMES/code-review.md:59]:
```suggestion
* Request a review from a Tech Lead if necessary.
* Request a review from a Product Manager if necessary.
```

File: READMES/code-review.md [hunk 11]
```diff
@@ -1,15 +1,99 @@
...
+* Request a review from a Product Owner on CI if necessary.
```
Comment (oksana-c) [inline READMES/code-review.md:60]:
```suggestion
* Request a review from a Product Owner on CI if necessary. Note: issues that require PO review will be tagged using `Needs PO review` label. PRs addressing issues that need PO review should also be labelled.
```

File: READMES/code-review.md [hunk 12]
```diff
@@ -1,15 +1,99 @@
...
+
+### Tech Lead
+
+The Tech Lead is accountable for the code quality and architecture practices of the CMS application. In respect to the code review process, the tech lead must be added as a PR Reviewer if a change introduced by a Pull Request contains an architectural change or a significant new feature.
+
+### Product Owner
+
+The Product Manager is a person on the CMS team who works with the Product Owner on defining feature requests and priorities. The Product Manager can participate in the Code Review process and be brought in to review changes on the CI environment as specified in User Story Acceptance Criteria. The Code Owner will keep the Product Manager up to date on status.
```
Comment (oksana-c) [inline READMES/code-review.md:68]:
```suggestion
### Product Owner

The Product Owner is a person who makes and prioritizes feature requests. This person is normally a part of VA/DEPO organization. The Product Owner is not directly involved in the PR Review process. The Product Manager makes a decision of whether the Product Owner should be brought in to review changes on the CI environment.

### Product Manager

The Product Manager is a person on the CMS team who works with the Product Owner on defining feature requests and priorities. The Product Manager can participate in the PR Review process and be brought in to review changes on the CI environment as specified in User Story Acceptance Criteria. The Code Owner will keep the Product Manager up to date on status.
```

File: READMES/code-review.md [hunk 13]
```diff
@@ -1,15 +1,99 @@
...
+
+## PR Review process
+
+1. A Pull Request is created in Github as a `draft`.
+1. The PR Owner is set as the Assignee of the Pull Request.
+1. Add Code/Product team label.
```
Comment (oksana-c) [inline READMES/code-review.md:74]:
```suggestion
1. Add `Core Application Team` or `Product Team Support` label.
```

File: READMES/code-review.md [hunk 14]
```diff
@@ -1,15 +1,99 @@
...
+
+The Tech Lead is accountable for the code quality and architecture practices of the CMS application. In respect to the code review process, the tech lead must be added as a PR Reviewer if a change introduced by a Pull Request contains an architectural change or a significant new feature.
+
+### Product Owner
+
+The Product Manager is a person on the CMS team who works with the Product Owner on defining feature requests and priorities. The Product Manager can participate in the Code Review process and be brought in to review changes on the CI environment as specified in User Story Acceptance Criteria. The Code Owner will keep the Product Manager up to date on status.
+
+## PR Review process
+
+1. A Pull Request is created in Github as a `draft`.
```
Comment (oksana-c) [inline READMES/code-review.md:72]:
```suggestion
1. A Pull Request is created in Github as a `Draft`.
```

File: READMES/code-review.md [hunk 15]
```diff
@@ -1,15 +1,99 @@
...
+1. The PR Owner is set as the Assignee of the Pull Request.
+1. Add Code/Product team label.
+1. The PR Owner will test the changes in CI. When the PR Owner is confident the changes are ready for review, the PR Owner will set the Pull Request to `Ready for Review`.
+1. If appropriate, the PR Owner will add one or multiple PR Reviews as reviewers on the Pull Request.
```
Comment (oksana-c) [inline READMES/code-review.md:76]:
```suggestion
1. PR Owner will request one or multiple PR Reviewers for the Pull Request as needed.
```

File: READMES/code-review.md [hunk 16]
```diff
@@ -1,15 +1,99 @@
...
+1. If a change introduced by a Pull Request introduces an architectural change or a significant new feature, the Tech Lead will take the role as the PR Reviewer and a reviewers on the Pull Request.
```
Comment (oksana-c) [inline READMES/code-review.md:77]:
```suggestion
1. If a change introduced by a Pull Request introduces an architectural change or a significant new feature, the Tech Lead will take the role of the PR Reviewer on the Pull Request.
```

File: READMES/code-review.md [hunk 17]
```diff
@@ -1,15 +1,99 @@
...
+1. PR Reviewer will look at the code and make comments/suggestions.
+1. PR Reviewer will test the code in the CI environment to make sure the request features are implemented and no other regressions exist.
+1. PR Reviewer will provide feedback on the Pull Request using the Review Changes feature. The feedback should be in a single review and not several individual comments on the Pull Request.
```
Comment (oksana-c) [inline READMES/code-review.md:80]:
```suggestion
1. PR Reviewer will provide feedback on the Pull Request using the Review Changes feature. Good feedback practice is to leave a single review and not several individual comments on the Pull Request.
```

File: READMES/code-review.md [hunk 18]
```diff
@@ -1,15 +1,99 @@
...
+1. The PR Owner will provide changes requested by the Code Review and answer all questions.
+1. When PR Reviewer is satisfied the changes introduced by the Pull Request are of high quality, address the related issue and will not introduce new regressions, the Pull Request is approved.
```
Comment (oksana-c) [inline READMES/code-review.md:82]:
```suggestion
1. When PR Reviewer has identified that the changes introduced by the Pull Request are of high quality, address the related issue and will not introduce new regressions, the Pull Request is approved.
```

File: READMES/code-review.md [hunk 19]
```diff
@@ -1,15 +1,99 @@
...
+The following automated code tools must pass before a Pull Request can be merged.
+
+* PHPCS
+* PHPMD
+
+## Definition of Done
+
+* Automated Tests Pass
+* Code Quality Tools pass
+* Manual Code Review Approved
```
Comment (oksana-c) [inline READMES/code-review.md:96]:
```suggestion
* Manual Code Review Approved
* Acceptance Criteria in related issue are met
```

**Review by oksana-c — CHANGES_REQUESTED** [review]

**Comment by indytechcook** [comment]
> One question: why is it important for the PR to be assigned to the author? I'm guessing that serves some function, I'm just not sure what. And maybe that should be in the document?

@ndouglas it's important to know who the code owner is both as a reviewer and as an owner.  Like if I have 5 PR's I'm owning I can filter the list down quickly.

File: READMES/code-review.md (hunk 1 above)
Comment (indytechcook) [reply READMES/code-review.md:89]: I'd like to separate out tests from code quality.  Eventually moving to something like https://codeclimate.com/ or https://www.exakat.io/en/.  Tugboat also provides the ability to fail the build before tests even start on code quality fails.

File: READMES/code-review.md (hunk 9 above)
Comment (indytechcook) [reply READMES/code-review.md:54]: I meant to delete this line.  We don't need a spec as automated tests, code quality and manual review is already listed on this doc

File: READMES/code-review.md (hunk 11 above)
Comment (indytechcook) [reply READMES/code-review.md:60]: Reminder for me to update the PR template in github

File: READMES/code-review.md [hunk 20]
```diff
@@ -1,15 +1,101 @@
...
+### PR Reviewer
+
+A PR Review is one or more people responsible for reviewing all the changes a Pull Request introduces. There can be multiple PR Reviewers per PR. A PR Reviewer can be anyone that is part of the department-of-veterans-affairs/va.gov-cms group. If a change introduced by a Pull Request introduces an architectural change or a significant new feature, a Tech Lead and a Product Manager (as needed) should be the PR Reviewer.
+
+* A PR Reviewer will self assign or be assigned by the PR Owner as a Reviewer on the Pull Request.
+* PR Reviewer Responsibilities
+* Review code for quality passed upon Insert Spec Here?
+* Test Code changes on CI
+* Document what was tested providing screenshots if necessary.
+* Follow up with questions/feedback PR Code Owner in a timely manner
```
Comment (acrollet) [inline READMES/code-review.md:53]:
```suggestion
* Follow up with questions/feedback for the PR Code Owner in a timely manner
```

File: READMES/code-review.md [hunk 21]
```diff
@@ -1,15 +1,101 @@
...
+* Merging the Pull Request when it's approved, tests are passing and the PR meets a Definition of Done.
+* Validating the changes in the Pull Request fulfill the needs of the Product Owner and ACs in related issue
+* Communicating any status about the deploy process, extra downtime, etc to the appropriate team members and Product Owner.
+* Provide notes on how to test and any necessary screenshots of changes.
+* Request a review from a Tech Lead if necessary.
+* Request a review from a Product Owner on CI if necessary.
+
+### PR Reviewer
+
+A PR Review is one or more people responsible for reviewing all the changes a Pull Request introduces. There can be multiple PR Reviewers per PR. A PR Reviewer can be anyone that is part of the department-of-veterans-affairs/va.gov-cms group. If a change introduced by a Pull Request introduces an architectural change or a significant new feature, a Tech Lead and a Product Manager (as needed) should be the PR Reviewer.
```
Comment (acrollet) [inline READMES/code-review.md:46]:
```suggestion
A PR Reviewer is one or more person(s) responsible for reviewing all the changes a Pull Request introduces. There can be multiple PR Reviewers per PR. A PR Reviewer can be anyone that is part of the department-of-veterans-affairs/va.gov-cms group. If a change introduced by a Pull Request introduces an architectural change or a significant new feature, a Tech Lead and a Product Manager (as needed) should be the PR Reviewer.
```

File: READMES/code-review.md [hunk 22]
```diff
@@ -1,15 +1,101 @@
...
+* Request a review from a Product Manager if necessary.
+* Request a review from a Product Owner on CI if necessary. Note: issues that require PO review will be tagged using `Needs PO review` label. PRs addressing issues that need PO review should also be labelled.
+
+### Tech Lead
+
+The Tech Lead is accountable for the code quality and architecture practices of the CMS application. In respect to the code review process, the tech lead must be added as a PR Reviewer if a change introduced by a Pull Request contains an architectural change or a significant new feature.
+
+### Product Owner
+
+The Product Owner is a person who makes and prioritizes feature requests. This person is normally a part of VA/DEPO organization. The Product Owner is not directly involved in the PR Review process. The Product Manager makes a decision of whether the Product Owner should be brought in to review changes on the CI environment.
```
Comment (acrollet) [inline READMES/code-review.md:65]:
```suggestion
The Product Owner is a person who makes and prioritizes feature requests. This person is normally a part of VA/DEPO organization. The Product Owner is not directly involved in the PR Review process. The Product Manager decides whether the Product Owner should be brought in to review changes on the CI environment.
```

**Review by ethanteague — DISMISSED** [review]

**Comment by indytechcook** [comment]
I listed a few possible follow ups after a discussion in slack: https://github.com/department-of-veterans-affairs/va.gov-cms/issues/3157#issuecomment-726805558.

**Pull request merged**